shared/subspace-data-retrieval/src/object_fetcher.rs

```diff
@@ -0,0 +1,456 @@
+                    progress > offset_in_segment
+                })
+                .ok_or_else(|| {
+                    error!(
+                        offset_in_segment,
+                        %segment_index,
+                        ?mapping,
+                        "Failed to find item at offset in segment"
+                    );
+
+                    Error::NoSegmentItem(offset_in_segment, segment_index, mapping)
+                })?;
+
+            match segment_item {
+                SegmentItem::Block { bytes, .. }
+                | SegmentItem::BlockStart { bytes, .. }
+                | SegmentItem::BlockContinuation { bytes, .. } => {
+                    // Rewind back progress to the beginning of the number of bytes
+                    progress -= bytes.len();
+                    // Get a chunk of the bytes starting at the position we care about
+                    Vec::from(&bytes[offset_in_segment - progress..])
+                }
+                segment_item => {
+                    // TODO: create a Display impl for SegmentItem that is shorter than the entire
+                    // data contained in it
+                    error!(
+                        ?segment_item,
+                        offset_in_segment,
+                        %segment_index,
+                        ?mapping,
+                        "Unexpected segment item",
+                    );
+
+                    return Err(Error::UnexpectedSegmentItem(
+                        offset_in_segment,
+                        segment_index,
+                        format!("{segment_item:?}"),
+                        mapping,
+                    ));
+                }
+            }
+        };
+
+        // Return an error if the length is unreasonably large, before we get the next segment
+        if Self::decode_data_length(&mut data.as_slice(), &mapping)?.is_some() {
+            if let Ok(data) = Vec::<u8>::decode(&mut data.as_slice()) {
+                // We have the full object
+                return Ok(data);
+            }
+        }
+
+        // We need to read an extra segment to get the object length, or the full object.
+        // Segments are much larger than MAX_OBJECT_LENGTH, so we won't need to read multiple segments.
+        // TODO: what if the next segment has no BlockContinuations?
+        let Segment::V0 { items } = self.read_segment(segment_index + SegmentIndex::ONE).await?;
+        for segment_item in items {
+            if let SegmentItem::BlockContinuation { bytes, .. } = segment_item {
+                data.extend_from_slice(&bytes);
+
+                // We already downloaded the data, and there are no more segments to download, so
+                // we don't check the length again. In rare cases this can allow oversized objects
+                if let Ok(data) = Vec::<u8>::decode(&mut data.as_slice()) {
+                    return Ok(data);
+                }
+            }
```

Comment:
List all segment item variants exhaustively, ignore most, match `SegmentItem::Block { bytes, .. } | SegmentItem::BlockStart { bytes, .. }` and error if we see them.